### ISSUE TITLE
Style Features

### ISSUE THREAD

**Issue by dtdesign** [WoltLab/WCF#2848]
The default style is pretty flexible, but as with all things, even if there aren't a lot of limits, it becomes quite tedious when different styles all need to implement the same feature themselves. Examples include a different order of the logo/menu area or a layout where the entire page is somewhat framed.

Enter the realm of style features, a toggle for styles that alters the system's CSS to generate a different set of style rules on demand. These style features are enabled per style and require the compatibility with the style API 5.2, they're also available for tainted styles only, effectively preventing end users from accidentially toggling these options for unspported styles created by third parties.

* [ ] Experimental feature that cannot be extended by plugins.
* [ ] New Feature **Sticky menu below the logo** (desktop only)
  * [ ] Reversed order of the `.pageHeaderPanel` and `.pageHeaderFacade` containers.
  * [ ] The menu is static unless it is scrolled out of view in which case it becomes fixed to the window.
* [ ] New Feature **Page Frame**
  * The width restrictions currently enforced by `.layoutBoundary` is applied to `.pageContainer` instead.
  * `.layoutBoundary` will continue to maintain its padding for visual purposes.
  * Might require new color inputs for the overall site background.
* [ ] Optional search icon in the page header, support for an always visible search bar.

Existing issues:
* [ ] The comment system uses the generic `containerList` element which renders a superfluous border when there are no comments. Might be only solved by updating the JS code.

**Comment by jens1o** [comment]
>     * Experimental feature that cannot be extended by plugins.

Why do you want to avoid this functionality getting extended by plug-ins?

**Comment by dtdesign** [comment]
It's designed to toggle parts of the core CSS, for example, preventing some CSS from being generated at all. This is diffierent to the "cascading" aspect of CSS where you simply declare styles that overrule existing ones, because these style features will change the core CSS itself.

This is not to be confused with styles that offer custom SCSS variables to toggle the behavior of their style. We're not talking about "userland" code!

In short, these CSS features are intended to give us the tools to provide different style behaviors without breaking the compatibility with existing styles. We need to be extra carefully with any change we make and especially we need to be aware of any consequences that our changes have.

**Comment by jens1o** [comment]
Thanks for the clarification. Sounds reasonable.

**Comment by KittMedia** [comment]
Why should you reverse order of the `.pageHeaderPanel` and `.pageHeaderFacade` containers. in order to achieve the first? This can be done with CSS and thus without changing the templates (in an unnecessary way).

**Comment by dtdesign** [comment]
This is intended to be purely a CSS based change by utilizing the `order` property of flex box. Maybe even using CSS' grid, depending on what proves to be the most stable approach.

**Comment by KittMedia** [comment]
Ah sorry, misinterpreted on my end. 🙈

**Comment by MatzeKitt** [comment]
> Might require new color inputs for the overall site background.

Please, no. We already have content and content container background colors, which already are usually not necessary for the most custom designs. An additional color would add additional garbage to this.

**Comment by dtdesign** [comment]
The more I look into this issue and its style related siblings, the less I am convinced that this is possible with the current style system. The use of CSS preprocessors dates back to a time where CSS variables where nothing but a shallow dream in the far, far future. The use of flexbox is ubiquitous, even for pure layout purposes that are much better solved using CSS grids today, for example, the awful `.pageHeader` mess.

It feels a bit like sitting in between chairs, the current style system with its (S)CSS and templates has matured over years and a lot of custom styles exist that build upon it. Any significant change will break the compatibility, because there is too much set in stone already. This extends deep into the realm of accessibility considering the common misuse of `ol` / `ul` and the overall inconsistent a11y behavior of interface components caused by multiple parallel implementations.

For now, we will most likely stick to what we have, the price of a significant overhaul somewhat outweighs the benefits. Just because something is inherently superior doesn't  mean it is actually better.

**Comment by MatzeKitt** [comment]
I don’t talk about compatibility, though. I talk about settings that are useless for the majority of the styles, which in my opinion should be avoided. It already wasn’t nice to remove the content container from all of my styles when it were introduced. And it was just for the purpose of the base style. Even many new styles just don’t use it this way.

**Comment by dtdesign** [comment]
This is entirey about compatibility. Maintaining compatibility means that we can only make subtle changes in a very narrow path, because everything beyond that will drive third party styles into the ground.

For example, we are effectively unable to innovate on the default style, because it _will_ cause incompatibilities. Any CSS change we make must be either closely inspected to not break something or be hidden behind a version guard that prevents these from affecting styles.

We have thought about introducing a new default style that is built on top of the existing one, similar to how we had the "Blue Sunrise" and the "WoltLab Basic" style back in the WCF 1.x era. However, it only provides the illusion of "freedom", because new components will still exist within the boundaries of the existing CSS. Also it means that the "basic" style will be more and more left behind as we saw with the step from WCF 1.0 to 1.1. Taking into account the current pacing with new releases, these discrepancies will grow even more. It has a reason why we dropped the distinction of a "basic" and a (fancy) "default" style, because it does not solve problems, but instead burries them beneath the surface.

So yes, this is about compatibility and has always been. The attempt at the content container you've mentioned yourself is a perfect example of the fundamental problem with compatibility. And oh boy, don't get me started on styles that ship with modified templates, that is a whole different level of madness and is among the most significant design flaws of the style system.

**Comment by MatzeKitt** [comment]
But then I really have to ask if the current way is the correct one? Change a known broking system to overcome some minor problems (at least for me, since the current implementation works, even if it’s not the best one).

What I want to say is: Please don’t change some minor things, which requires many style creators to become active in order to properly support the newest WoltLab Suite. If you’re not happy with the current system, get rid of it. And if you don’t want to do that in a new minor release (which is totally fine and reasonable), please just don’t change it.

At least I can see no real benefit from these changes that are reasonable to update all my styles. And yes, I know there is a compatibility layer. But if you want your styles to be marked as compatible with the current version you **have to** adapt the updates. Everything else would be a flaw to the customers.

**Comment by dtdesign** [comment]
Well, yes. You basically summed up my reasons for constantly pushing this issue back. You are aware that this was originally planned for 5.2 and is currently set to 5.5 for re-evaluation?

**Comment by MatzeKitt** [comment]
I did just mention that is was set to 5.4 before. I just wanted to point out that in my opinion it’s nothing that should be done before re-structuring the style system (hopefully with 6.0.0 😉).
